crates/types/src/traits/signature_key.rs

```diff
@@ -336,7 +336,10 @@ fn aggregate_fee_data<Metadata: EncodeBytes>(
     let mut fee_info = Vec::new();
     fee_info.extend_from_slice(fee_amount.to_be_bytes().as_ref());
     fee_info.extend_from_slice(metadata.encode().as_ref());
-    fee_info.extend_from_slice(vid_commitment.as_ref());
+    fee_info.extend_from_slice(
+        &bincode::serialize(&vid_commitment)
```

Comment:
It is possible. But `AsRef` is not something jellyfish promises in trait bounds. Hotshot code had to break abstraction in order to use it. (ie. Hotshot code relied on underlying implementation details of ADVZ commitment to access `as_ref`.)

Supporting `AsRef` is troublesome because it returns a *reference*, which basically forces the underlying implementation to store its data as a byte slice, as anything else would require a conversion to `[u8]` that immediately goes out of scope.

`AsRef` isn't really meant for serialization, which is what we actually want here. Instead, `AsRef` is intended only for "[cheap reference-to-reference conversion](https://doc.rust-lang.org/std/convert/trait.AsRef.html)".

## Options
1. Bite the bullet and impose a `AsRef` requirement. If we do this then ultimately we should do it upstream, ie. add `AsRef` to the trait bounds promised by jellyfish. If we choose this option then I'm willing to restore support for `AsRef` in this PR in anticipation of a future change to jellyfish API.
2. Add a helper function to hotshot that serializes a `VidCommitment` into a `Vec<u8>`. This removes the need for the ugly `bincode::serialize(...).expect(...)` everywhere. It's still there, of course. It just appears in one place instead of everywhere. Naturally, this option is slower because bincode needs to allocate heap memory and copy into it.

Reply:
Oh, I think I understand the ownership issue with `AsRef`.

I guess I'm less concerned about the cost of the copy (it's only a few hundred bytes I think?) as I am about the `bincode` and the `expect`. Is it possible for jellyfish to add an extra trait bound that exports an infallible `VidCommitment -> Vec<u8>`?

Reply:
Well, a lot of `Vec` operations could panic, too. I don't see any reason why panic via bincode is more likely than panic via Vec. bincode is fallible because it's extremely general and serde-compatible. In our application I expect it never to fail, so panic-on-failure is probably an acceptable solution. I confess I don't like the panic either, but that seems to be the nature of Rust.

If we were to add an infallible conversion to `Vec<u8>` then we'd be duplicating the functionality of serde. (Indeed, I would probably implement this "infallible" conversion via `bincode` + `expect` just to ensure compatibility with serde. 😛 )

Bottom line: if your only objection to bincode is that it might panic then I wouldn't worry about it. TBH my main objection is performance. Seems like we can't avoid touching the heap and copying bytes each time we access the commitment. It hurts my OCD not to break abstraction and return a reference to internal bytes. 😬

Reply:
My suggestion: add to `types/src/vid.rs` a `pub fn` to wrap the ugly `bincode` + `expect`. We could call it `as_bytes` or even `bincode_serialize`.

Reply:
On further consideration, we decided to restore use of `AsRef`. On balance, the benefits (ergonomics, efficiency) outweigh the cost of lost abstraction. Done in https://github.com/EspressoSystems/HotShot/pull/3794/commits/d74cc908174ab3ff2a824d5d29111ca9098620c4